pkg/store/sqlite/search.go

```diff
@@ -164,3 +172,105 @@ func NormURL(query string) string {
 
 	return ""
 }
+
+//===========================================================================
+// Ranking Search Results
+//===========================================================================
+
+type SearchRank struct {
+	query string
+	limit int
+	items RankList
+	idset map[ulid.ULID]struct{}
+}
+
+func NewSearchRank(query string, limit int) *SearchRank {
+	return &SearchRank{
+		query: query,
+		limit: limit,
+		items: make(RankList, 0, limit+1),
+		idset: make(map[ulid.ULID]struct{}, limit+1),
+	}
+}
+
+func (s *SearchRank) Add(id ulid.ULID, name string) bool {
+	return s.Append(&RankItem{ID: id, Name: name})
+}
+
+func (s *SearchRank) Append(item *RankItem) bool {
+	if _, ok := s.idset[item.ID]; ok {
+		return false // already in the list
+	}
+
+	item.Distance = fuzzy.RankMatchNormalizedFold(s.query, item.Name)
+	if item.Distance < 0 {
+		return false // not a match
+	}
+
+	item.Similarity = s.similarity(item.Distance, item.Name)
+	if item.Similarity < threshold {
+		return false // not within the similarity threshold
+	}
+
+	s.idset[item.ID] = struct{}{}
+	s.items = append(s.items, *item)
+
+	if len(s.items) > s.limit {
+		// Sort the items by distance to get the closest matches first
+		sort.Sort(s.items)
+
+		// Remove any IDs from any items that are not in the limit
+		for _, item := range s.items[s.limit:] {
+			delete(s.idset, item.ID)
+		}
+
+		// Trim the list to the limit
+		s.items = s.items[:s.limit]
+	}
+
+	return true
+}
+
+func (s *SearchRank) similarity(d int, term string) float64 {
+	if len(term) > len(s.query) {
+		return 1.0 - float64(d)/float64(len(term))
+	} else {
+		return 1.0 - float64(d)/float64(len(s.query))
+	}
+}
```

Comment:
This is just an approximate metric since I don't actually know the length of the term or the query w.r.t. to the distance metric (e.g. those strings are not fold normalized to encapsulate unicode and case-insensitivity); however this approximation is probably sufficient though some exact matches for unicode strings might end up being 0.99 or similar.

Reply:
Reference for similarity computation: https://www.baeldung.com/cs/string-similarity-edit-distance#levenshtein-distance